lib/ide/utils/src/handle.rs

```diff
@@ -0,0 +1,268 @@
+//! This module implements utilities that allow for a safer usage pattern of
+//! a shared data stored under `RefCell`. The primary goal is to allow accessing
+//! the data without any direct calls to `borrow_mut`.
+//!
+//! Any type that implements `IsWeakHandle` or `IsStrongHandle` (types provided
+//! out-of-the-box are Weak<RefCell<T>> and Strong<RefCell<T>> responsively)
+//! gets an extension method names `with`. It allows accessing a data through
+//! callback function that will get &mut access to the `T`.
+//!
+//! The safety is ensured by the execution model. The `with` function returns
+//! a `Future` that yields a result of callback. Because `borrow_mut` only
+//! applied in the `Future` implementation (being called by the executor), it
+//! is guaranteed to be safe (as executors do not allow self-entry and callback
+//! must be a synchronous function).
+
+use crate::prelude::*;
+
+use std::future::Future;
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+/// Packs given value into a Rc<RefCell<T>>.
+pub fn strong<T>(t:T) -> StrongHandle<T> { Rc::new(RefCell::new(t)) }
+
+/// A strong (shared ownership) handle to some internally mutable data.
+pub type StrongHandle<T> = Rc<RefCell<T>>;
+
+/// A weak handle to some internally mutable data.
+pub type WeakHandle<T> = Weak<RefCell<T>>;
+
```

Comment:
If this is meant to be a safe interface I dont like the idea to still allow me to `borrow_mut` things here. Why not newtypes?

Reply:
+1

Reply:
done